java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java

```diff
@@ -167,4 +185,57 @@ protected void process(
       throw new IOException(e);
     }
   }
+
+  private static boolean IsKeyPresentIn(List<String> keys, String key) {
+    return keys == null || keys.isEmpty() || keys.contains(key);
+  }
+
+  private List<Map<String, Object>> getNodesInfo() {
+    List<RemoteProxy> proxies = getRegistry().getAllProxies().getSorted();
+    return proxies.stream().map(this::getNodeInfo).collect(toList());
+  }
+
+  private Map<String, Object> getNodeInfo(RemoteProxy remoteProxy) {
+    return ImmutableSortedMap.of(
+        "Id", remoteProxy.getId(),
+        "browsers", getInfoFromAllSlotsInNode(remoteProxy.getTestSlots())
+    );
+  }
+
+  private List<Map<String, Object>> getInfoFromAllSlotsInNode(List<TestSlot> slots) {
+    List<Map<String, Object>> browsers = Lists.newArrayList();
+    Map<String, List<TestSlot>>
+        slotsInfo = slots.stream().collect(groupingBy(HubStatusServlet::getBrowser));
+    for (Map.Entry<String, List<TestSlot>> each : slotsInfo.entrySet()) {
+      String key = each.getKey();
+      Map<String, Object> value = getSlotInfoPerBrowserFlavor(each.getValue());
+      browsers.add(ImmutableSortedMap.of("browser", key, "slots", value));
+    }
+    return browsers;
+  }
+
+  private Map<String, Object> getSlotInfoPerBrowserFlavor(List<TestSlot> slots) {
+    Map<String, Integer> byStatus = slots.stream().collect(groupingBy(this::status, counting()));
+    int busy = byStatus.computeIfAbsent(BUSY, status -> 0);
+    int free = byStatus.computeIfAbsent(FREE, status -> 0);
+    int total = busy + free;
+
+    return ImmutableSortedMap.of(TOTAL, total, BUSY, busy);
+  }
+
+  private String status(TestSlot slot) {
+    if (slot.getSession() == null) {
+      return FREE;
+    }
+    return BUSY;
+  }
+
+  private static String getBrowser(TestSlot slot) {
+    return slot.getCapabilities().get(CapabilityType.BROWSER_NAME).toString();
+  }
+
+  private static <T> Collector<T, ?, Integer> counting() {
+    return reducing(0, e -> 1, Integer::sum);
```

Comment:
This looks like an IntStreams `sum` method, but presumably that isn't appropriate to use here